src/charm.py

```diff
@@ -31,12 +31,62 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.get_primary_action,
+                               self._on_get_primary_action)
+
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self._number_of_expected_units == self._number_of_current_units:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
+                str(self._number_of_expected_units),
+                str(self._number_of_current_units),
+            )
+            return
+
+        # do not initilise replica set until all peers have started mongod
+        for peer in self._peers.units:
+            mongo_peer = self._single_mongo_replica(
+                str(self._peers.data[peer].get("private-address")))
+            if not mongo_peer.is_ready(as_stand_alone=True):
+                logger.debug(
+                    "unit: %s is not ready, cannot initilise replica set "
+                    + "until all units are ready, deferring on relation-joined",
+                    peer,
+                )
+                event.defer()
```

Comment:
I would be hesitant to defer a relation joined event. Since when new units join you are bound to get more relation joined events. Uptil now the only use case I have come across where a defer becomes essential is the start event.